serving/docker/partition/run_multi_node_setup.py

```diff
@@ -0,0 +1,267 @@
+        """
+        Downloads model from s3 given s3 URL using s5cmd
+        """
+        logger.info("Downloading model...")
+        if not model_url or not model_url.startswith("s3://"):
+            logger.warn(
+                f"Model {model_id} to be downloaded is not s3. Will attempt to download from HF hub."
+            )
+            return
+
+        download_dir = os.environ.get("SERVING_DOWNLOAD_DIR", "")
+
+        if not download_dir:
+            download_dir = Path(
+                self.get_cache_dir()) / "download" / self.hash(model_url)
+            download_dir.mkdir(parents=True, exist_ok=True)
+
+        if Path("/opt/djl/bin/s5cmd").is_file():
+            if not model_url.endswith("*"):
+                if model_url.endswith("/"):
+                    model_url = model_url + "*"
+                else:
+                    model_url = model_url + "/*"
+
+            commands = [
+                "/opt/djl/bin/s5cmd", "--retry-count", "1", "sync", model_url,
+                download_dir
+            ]
+        else:
+            commands = ["aws", "s3", "sync", model_url, download_dir]
+
+        subprocess.run(commands)
+
+        if not glob.glob(os.path.join(download_dir, "*")):
+            raise Exception("Model download from s3url failed")
+
+        logging.info(f"Model download complete.")
+        return True
+
+    @retry(stop_max_delay=60000, wait_fixed=2000)
+    def get_and_write_cluster_ssh_key(self):
+        """
+        Gets ssh key generated by the leader node, and writes it to authorized keys.
+        Private key may be necessary for bi-directional communication back to leader node, when using mpirun
+        """
+        logger.info(f"Querying ssh key from leader node")
+
+        endpoint_url = self.endpoint_url + "sshpublickey"
+
+        try:
+            response = requests.get(endpoint_url)
+
+            if response.status_code == 200:
+                file_data = response.content
+
+                home_dir = Path.home()
+                ssh_dir = home_dir / ".ssh"
+                ssh_dir.mkdir(mode=0o700, parents=True, exist_ok=True)
+
+                id_rsa_pub_path = ssh_dir / "id_rsa.pub"
+
+                with open(id_rsa_pub_path, "wb") as file:
+                    file.write(file_data)
+
+                id_rsa_pub_path.chmod(0o644)
+
+                authorized_keys_path = ssh_dir / "authorized_keys"
+
+                with open(authorized_keys_path, "wb") as file:
+                    file.write(file_data)
+
+                authorized_keys_path.chmod(0o644)
+
+                logger.info(f"SSH key written to {id_rsa_pub_path}")
+                logger.info(f"SSH key written to {authorized_keys_path}")
+            else:
+                logger.info("Failed to fetch data from the endpoint.")
+        except Exception as e:
+            logger.error(f"An error occurred: {e}")
+
+        endpoint_url = self.endpoint_url + "sshprivatekey"
```

Comment:
Do we really need private key? does mpirun use bi-directional ssh?

Reply:
Removed it all, as bi-directional communication does not seem to be needed currently.